enhancements/machine-api/cluster-autoscaler-integration.md

```diff
@@ -0,0 +1,456 @@
+denote the maximum and minimum sizes that the autoscaler should scale the Node group within, eg:
+```yaml
+metadata:
+  annotations:
+    - "machine.openshift.io/cluster-api-autoscaler-node-group-max-size": "6"
+    - "machine.openshift.io/cluster-api-autoscaler-node-group-min-size": "1"
+```
+
+### User Stories
+
+#### Story 1
+
+As a user of OpenShift, I would like to be able to autoscale workloads on my cluster as required by my users, and have
+the cluster automatically expand to meet the new capacity requirements
+
+#### Story 2
+
+As a user of OpenShift, if the workloads on my cluster reduce leaving unused capcity, I would like my cluster to
+automatically scale back to reduce cost and wasted resources.
+
+#### Story 3
+
+As a developer of OpenShift, I would like an agnostic integration that specifies a minimal API contract for autoscaling
+so that it is compatible with multiple versions and implementations of the machine API simultaneously to reduce the
+burden of fork/extend style development.
+
+### Implementation Details
+
+This integration will satisfy the [CloudProvider](https://github.com/kubernetes/autoscaler/blob/a00bf59159bda6d70e86741764802d11e7d253f4/cluster-autoscaler/cloudprovider/cloud_provider.go#L48-L91)
+interface by mapping the [NodeGroup](https://github.com/kubernetes/autoscaler/blob/a00bf59159bda6d70e86741764802d11e7d253f4/cluster-autoscaler/cloudprovider/cloud_provider.go#L103-L170)
+concept to MachineSets and MachineDeployments that opt-in to autoscaling.
+
+#### Opting in to Autoscaling
+
+In order for a MachineSet or MachineDeployment to be considered for autoscaling, the autoscaler integration will need
+autoscaling bounds for the NodeGroup, ie. it will need to know the maximum and minimum size that the MachineSet/
+MachineDeployment should be.
+
+This information will be provided to the Cluster Autoscaler by way of annotations on the MachineSet/MachineDeployment, eg:
+```yaml
+metadata:
+  annotations:
+    - "machine.openshift.io/cluster-api-autoscaler-node-group-max-size": "6"
+    - "machine.openshift.io/cluster-api-autoscaler-node-group-min-size": "1"
+```
+
+#### Provider
+
+The Provider will implement the [CloudProvider](https://github.com/kubernetes/autoscaler/blob/a00bf59159bda6d70e86741764802d11e7d253f4/cluster-autoscaler/cloudprovider/cloud_provider.go#L48-L91)
+interface that is required to integrate Machine API with the cluster autoscaler.
+
+It will essentially wrap the Machine Controller which is responsible for maintaining the state of the cluster from an
+autoscaling perspective.
+
+#### Machine Controller
+
+The Machine Controller will be responsible for coordinating the underlying MachineSets and MachineDeployments to provide
+a uniform interface to the core of the Cluster Autoscaler. It will be responsible for:
+- Maintaining a cache of the state of the target cluster using informers
```

Comment:
It would be more valuable to list which resource types it will watch. Seems like caching is an assumed behavior for a controller these days.